metrics/internal/collectors/storageconsumer.go

```diff
@@ -0,0 +1,123 @@
+	"k8s.io/klog"
+)
+
+var _ prometheus.Collector = &StorageConsumerCollector{}
+
+type StorageConsumerCollector struct {
+	Informer                cache.SharedIndexInformer
+	StorageConsumerMetadata *prometheus.Desc
+	AllowedNamespace        string
+}
+
+type StorageConsumerLister interface {
+	List(selector labels.Selector) (ret []*ocsv1alpha1.StorageConsumer, err error)
+}
+
+type storageConsumerLister struct {
+	indexer cache.Indexer
+}
+
+func NewStorageConsumerLister(indexer cache.Indexer) StorageConsumerLister {
+	return &storageConsumerLister{indexer: indexer}
+}
+
+func (s *storageConsumerLister) List(selector labels.Selector) (ret []*ocsv1alpha1.StorageConsumer, err error) {
+	err = cache.ListAll(s.indexer, selector, func(m interface{}) {
+		ret = append(ret, m.(*ocsv1alpha1.StorageConsumer))
+	})
+	return ret, err
+}
+
+func NewStorageConsumerCollector(opts *options.Options) *StorageConsumerCollector {
+	ocsClient, err := getocsClient(opts)
+	if err != nil {
+		klog.Error(err)
+	}
+	lw := cache.NewListWatchFromClient(ocsClient, "storageconsumers", metav1.NamespaceAll, fields.Everything())
+	sharedIndexInformer := cache.NewSharedIndexInformer(lw, &ocsv1alpha1.StorageConsumer{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	return &StorageConsumerCollector{
+		StorageConsumerMetadata: prometheus.NewDesc(
+			prometheus.BuildFQName("ocs", "storage_consumer", "metadata"),
+			`Attributes of OCS Storage Consumers`,
+			[]string{"storage_consumer_name", "capacity", "state", "granted_capacity"},
+			nil,
+		),
+		Informer: sharedIndexInformer,
+	}
+}
+
+func (c *StorageConsumerCollector) Collect(ch chan<- prometheus.Metric) {
+	storageConsumerLister := NewStorageConsumerLister(c.Informer.GetIndexer())
+	storageConsumers := getAllStorageConsumers(storageConsumerLister, c.AllowedNamespace)
+	if len(storageConsumers) > 0 {
+		c.collectStorageConsumersMetadata(storageConsumers, ch)
+	}
+}
+
+func (c *StorageConsumerCollector) Describe(ch chan<- *prometheus.Desc) {
+	ds := []*prometheus.Desc{
+		c.StorageConsumerMetadata,
+	}
+	for _, d := range ds {
+		ch <- d
+	}
+}
+
+func (c *StorageConsumerCollector) Run(stopCh <-chan struct{}) {
+	go c.Informer.Run(stopCh)
+}
+
+func (c *StorageConsumerCollector) collectStorageConsumersMetadata(storageConsumers []*ocsv1alpha1.StorageConsumer, ch chan<- prometheus.Metric) {
+	for _, storageConsumer := range storageConsumers {
+		ch <- prometheus.MustNewConstMetric(c.StorageConsumerMetadata,
+			prometheus.GaugeValue, 1,
+			storageConsumer.Name,
+			storageConsumer.Spec.Capacity.String(),
+			string(storageConsumer.Status.State),
+			storageConsumer.Status.GrantedCapacity.String())
+	}
+}
+
+func getAllStorageConsumers(lister StorageConsumerLister, namespace string) (storageConsumers []*ocsv1alpha1.StorageConsumer) {
+	storageConsumers, err := lister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("couldn't list StorageConsumer. %v", err)
```

Comment:
```suggestion
		klog.Errorf("couldn't list StorageConsumer. %v", err)
		return nil
```
As the `List()` function in the lister is capable to return a partial appended list (when an error occurs), please add an explicit `return nil` statement.

Reply:
Done

Reply:
Done